### PR TITLE
Add Support for calling executable

### PR DESCRIPTION
**Pull request by mario-vimal**

### CODE REVIEW

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 1]
```diff
@@ -91,13 +134,102 @@ PluggableAuthExternalAccountCredentials::
...
 void PluggableAuthExternalAccountCredentials::RetrieveSubjectToken(
-    HTTPRequestContext* ctx, const Options& /*options*/,
+    HTTPRequestContext* /*ctx*/, const Options& /*options*/,
     std::function<void(std::string, grpc_error_handle)> cb) {
   cb_ = cb;
-  // TODO: Execute command
+  if (output_file_path_ != "") {
+    std::string output_file_content = read_file_contents(output_file_path_);
+    if (output_file_content != "") {
+      OnRetrieveSubjectToken(output_file_content);
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: We need a little different logic here for the output file reading. If the output file is successful but expired, or shows an error message (but properly formatted) we should run the executable and use that. If the output file has malformed stuff we should throw an error

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 2]
```diff
@@ -91,13 +134,102 @@ PluggableAuthExternalAccountCredentials::
...
+        if (output_file_content_string != "") {
+          OnRetrieveSubjectToken(output_file_content_string);
+          return;
+        }
+      }
+      OnRetrieveSubjectToken(output);
+      return;
+    }
+  } else {
+    thr.detach();
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: I don't think we want to detach here, my understanding is this would allow the process to keep running independently, if we go over the timeout we should interrupt it - something like this: https://github.com/grpc/grpc/blob/2c81c5619e80d69c96496098879620ed973d7b2c/test/core/util/subprocess_posix.cc#L93C1-L97C2

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 3]
```diff
@@ -91,13 +134,102 @@ PluggableAuthExternalAccountCredentials::
...
+                          std::string*, std::string*)>
+      run_executable_task(run_executable);
+  std::future<bool> executable_output_future = run_executable_task.get_future();
+  std::string output, error;
+  std::thread thr(std::move(run_executable_task), subprocess, command_, envp,
+                  &output, &error);
+  if (executable_output_future.wait_for(std::chrono::seconds(
+          executable_timeout_ms_ / 1000)) != std::future_status::timeout) {
+    thr.join();
+    if (!executable_output_future.get()) {
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:293]: what does this check actually mean? That the executable didn't complete?

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 4]
```diff
@@ -91,13 +134,102 @@ PluggableAuthExternalAccountCredentials::
...
-  // TODO: Execute command
+  if (output_file_path_ != "") {
+    std::string output_file_content = read_file_contents(output_file_path_);
+    if (output_file_content != "") {
+      OnRetrieveSubjectToken(output_file_content);
+      return;
+    }
+  }
+
+  std::vector<std::string> envp = {
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: You may need to add the existing environment variables here, not sure exactly how its implemented for the exec method. In some other languages we needed to add in all of the current global env variables as well to this map.

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 5]
```diff
@@ -91,13 +134,102 @@ PluggableAuthExternalAccountCredentials::
...
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE=%d", 0),
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_IMPERSONATED_EMAIL=%s",
+                      get_impersonated_email(impersonation_url_)),
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE=%s",
+                      output_file_path_)};
+  Subprocess* subprocess = new Subprocess();
+  std::packaged_task<bool(Subprocess*, std::string, std::vector<std::string>,
+                          std::string*, std::string*)>
+      run_executable_task(run_executable);
+  std::future<bool> executable_output_future = run_executable_task.get_future();
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:286]: Will this all work with windows without having to change anything assuming the subprocess.communicate and subprocess.start are implemented for it?

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc (hunk 3 above)
Comment (mario-vimal) [reply src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:293]: get() method gives the return value of the Subprocess::Communicate() method. It returns true when the executable finished without crashing.

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc (hunk 5 above)
Comment (mario-vimal) [reply src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:286]: Most of the stuff in std:: should have implementations in both windows and linux. So hopefully this should work. But the file read needs windows specific code.

Comment (aeitzman) [reply src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:286]: Oh I think code to read the file already exists, https://github.com/grpc/grpc/blob/2c81c5619e80d69c96496098879620ed973d7b2c/src/core/lib/security/credentials/external/file_external_account_credentials.cc#L110C1-L111C62

from the file based credentials

Comment (mario-vimal) [reply src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:286]: Sweet, I'll use this.

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 6]
```diff
@@ -91,13 +153,182 @@ PluggableAuthExternalAccountCredentials::
...
+    if (!executable_output_future.get()) {
+      FinishRetrieveSubjectToken(
+          "",
+          GRPC_ERROR_CREATE(absl::StrFormat(
+              "Failed reading output from the executable: %s", error_string)));
+      return;
+    } else {
+      if (output_file_path_ != "") {
+        std::string output_file_content_string =
+            read_file_contents(output_file_path_);
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: nit: replace file read method here too and delete the custom method you added

File: src/core/lib/gpr/subprocess.cc [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+  std::vector<std::string> arg_vector = absl::StrSplit(command, " ");
+  char** argv = (char**)malloc(sizeof(char*) * (arg_vector.size() + 1));
+  for (int i = 0; i < arg_vector.size(); i++) {
+    argv[i] = portable_strdup(arg_vector[i].c_str());
+  }
+  argv[arg_vector.size()] = nullptr;
+  int32_t environ_count = 0, i = 0;
+  while (environ[environ_count] != nullptr) environ_count += 1;
+  char** envp =
+      (char**)malloc(sizeof(char*) * (environ_count + envp_vector.size() + 1));
```
Comment (aeitzman) [inline src/core/lib/gpr/subprocess.cc]: Can you put this logic in the pluggable auth class, and then have the subprocess just call with whatever environment variables are provided? We want to use the existing environ variables but not every use case would.

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 8]
```diff
@@ -17,20 +17,89 @@
...
 
 #define DEFAULT_EXECUTABLE_TIMEOUT_MS 30000  // 30 seconds
 #define MIN_EXECUTABLE_TIMEOUT_MS 5000       // 5 seconds
 #define MAX_EXECUTABLE_TIMEOUT_MS 120000     // 120 seconds
+#define SAML_SUBJECT_TOKEN_TYPE "urn:ietf:params:oauth:token-type:saml2"
+#define GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES \
+  "GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES"
+#define GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES_ACCEPTED_VALUE "1"
+
+char* portable_str_dup(const char* s) {
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: Don't think this is necessarily an issue, but just want to double check that these basic functions aren't present somewhere in the library already

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 9]
```diff
@@ -83,21 +144,200 @@ PluggableAuthExternalAccountCredentials::
...
+  if (isKeyPresent(executable_json, "output_file"))
+    output_file_path_ = getStringValue(executable_json, "output_file");
+}
+
+void PluggableAuthExternalAccountCredentials::CreateExecutableResponse(
+    std::string executable_output_string) {
+  auto executable_output = JsonParse(executable_output_string);
+  if (!executable_output.ok()) {
+    FinishRetrieveSubjectToken(
+        "", GRPC_ERROR_CREATE("The output_file specified contains an invalid "
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: this won't always be an output file right?

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 10]
```diff
@@ -83,21 +144,200 @@ PluggableAuthExternalAccountCredentials::
...
+      CreateExecutableResponse(std::string(output_file_content));
+      if (executable_response_.success &&
+          !isExpired(executable_response_.expiration_time)) {
+        FinishRetrieveSubjectToken(executable_response_.subject_token,
+                                   absl::OkStatus());
+        return;
+      }
+    }
+  }
+  std::vector<std::string> envp_vector = {
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:263]: nit: can you add a few comments explaining what each section of retrieveSubjectToken is doing

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 11]
```diff
@@ -83,21 +144,200 @@ PluggableAuthExternalAccountCredentials::
...
+  std::string output_string, error_string;
+  std::thread thr(std::move(run_executable_task), subprocess, command_, envp,
+                  &output_string, &error_string);
+  if (executable_output_future.wait_for(std::chrono::seconds(
+          executable_timeout_ms_ / 1000)) != std::future_status::timeout) {
+    thr.join();
+    if (!executable_output_future.get()) {
+      FinishRetrieveSubjectToken(
+          "", GRPC_ERROR_CREATE(absl::StrFormat(
+                  "Executable failed with error: %s", error_string)));
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:297]: Can we get the error code as well or just the stderror here?

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc [hunk 12]
```diff
@@ -83,21 +144,200 @@ PluggableAuthExternalAccountCredentials::
...
+      return;
+    }
+    FinishRetrieveSubjectToken(executable_response_.subject_token,
+                               absl::OkStatus());
+    return;
+  } else {
+    subprocess->KillChildSubprocess();
+    FinishRetrieveSubjectToken(
+        "", GRPC_ERROR_CREATE(
+                absl::StrFormat("Executable timeout exceeded %d milliseconds.",
```
Comment (aeitzman) [inline src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: nit: "The executable failed to finish within the timeout of %d milliseconds"

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc (hunk 8 above)
Comment (mario-vimal) [reply src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: Thanks for the tip, found a str_dup method already implemented, so I'll use that.
The other two methods have no implementation and it will be nice to have them in gpr/json_util. 
Should I do a PR for this?

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc (hunk 11 above)
Comment (mario-vimal) [reply src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc:297]: This case is when the executable crashes and doesn't produce output/output_file that are going to have error_code and error_message .

File: src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc (hunk 8 above)
Comment (aeitzman) [reply src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc]: yeah thats a great idea, that would be a good stand alone pr
